Apps/WebClient/src/Server/Services/UserEmailService.cs

```diff
@@ -58,38 +58,44 @@ public class UserEmailService : IUserEmailService
         }
 
         /// <inheritdoc />
-        public bool ValidateEmail(string hdid, Guid inviteKey, string bearerToken)
+        public PrimitiveRequestResult<bool> ValidateEmail(string hdid, Guid inviteKey, string bearerToken)
         {
             this.logger.LogTrace($"Validating email... {inviteKey}");
-            bool retVal = false;
+            PrimitiveRequestResult<bool> retVal = new PrimitiveRequestResult<bool>();
             MessagingVerification? emailInvite = this.messageVerificationDelegate.GetByInviteKey(inviteKey);
 
-            if (emailInvite != null &&
-                emailInvite.HdId == hdid &&
-                !emailInvite.Validated &&
-                !emailInvite.Deleted &&
-                emailInvite.VerificationAttempts < MaxVerificationAttempts &&
-                emailInvite.ExpireDate >= DateTime.UtcNow)
+            if (emailInvite?.HdId != hdid ||
+                emailInvite?.Validated == true ||
+                emailInvite?.Deleted == true)
+            {
+                // Invalid Verification Attempt
+                emailInvite = this.messageVerificationDelegate.GetLastForUser(hdid, MessagingVerificationType.Email);
+                if (emailInvite != null &&
+                    !emailInvite.Validated)
+                {
+                    emailInvite.VerificationAttempts++;
+                    this.messageVerificationDelegate.Update(emailInvite);
+                }
+                retVal.ResultStatus = ResultType.Error;
+            }
+            else if (emailInvite.VerificationAttempts >= MaxVerificationAttempts ||
+                     emailInvite.ExpireDate < DateTime.UtcNow)
+            {
+                // Verification Expired
+                retVal.ResultStatus = ResultType.ActionRequired;
```

Comment:
why don't we name VerificationExpired instead of ActionRequired? as ActionRequired may also mean that user has not click the link but the link is still valid right?

Reply:
ActionRequired is the type of error. No need to create a new type for every single error

Reply:
resulttypes are generic and used for common error handling and its not specific to the email verification